rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,117 @@
+---
+feature: nix-language-version
+start-date: 2022-12-12
+author: @fricklerhandwerk
+co-authors: @thufschmitt @Ericson2314
+shepherd-team: 
+shepherd-leader: 
+related-issues: https://github.com/NixOS/nix/issues/7255
+---
+
+# Summary
+[summary]: #summary
+
+Introduce a convention to determine which version of the Nix language grammar to use for parsing a Nix file.
+
+# Motivation
+[motivation]: #motivation
+
+Currently it's impossible to introduce backwards-incompatible changes to the Nix language without breaking existing setups.
```

Comment:
I'm answering @tazjin's [comment](https://github.com/NixOS/rfcs/pull/137#issuecomment-1355226412) in this thread

Here's some things I'd like to see deprecated/changed, but there's probably more: 
- The `with` expression since it's very ambiguous, see also https://github.com/NixOS/rfcs/pull/110
- `toString`, it does too many things, especially the list-to-string coercion is problematic
- Similarly, string coercion in derivations, which uses the same mechanism as `toString` underneath
- The `let { body` syntax
- Improving confusing path semantics, see https://github.com/NixOS/nix/issues/7338
- `__functor` and `__toString`, probably
- Making builtins more consistent, e.g. not exposing `map`, `removeAttrs`, `fetchGit` and a bunch more on the global scope

Reply:
A lot of those are sensible (and I'd probably add a few more things, esp. weird scoping related behaviour/bugs), but that's a limited set and it doesn't justify (to me) reducing the barrier to making a language change by implicitly accepting them as a common occurence.

As all of these are more like cleanup of existing misbehaviours, it seems more reasonable to me to start warning about those behaviours (as tvix already does for some of them!) and then plan a future cutoff with a bump of `builtins.langVersion` that disable them.

Reply:
Oh yeah I haven't listed any of the new features I'd like to see:
- Proper destructuring support for attribute sets (also in `let` bindings)
- Being able to reassign variable names while destructuring (something like `{ "a" = first, "b" = second }: first + second`)
- A syntax to index into lists, e.g. `[ 1 2 3 ].0 == 1`
- Using `,` to delimit elements of a list expression, like `[ 1, 2, 3 ]`
- Doing something about `?` meaning two different things depending on where it occurs (`{ x ? "" }:` vs `x ? ""`)
- Proper support for importing and exporting Nix code, some builtin module system, especially also to enable better auto-completion (big discussion)
- More builtins in general (debatable whether that's part of the language, `builtins.langVersion` at least wasn't intended for `bulitins` additions, since you can just check with `builtins ? foo` whether a specific one is available)

While yes, every list of things we'd like to see changed is finite and might eventually be implemented, being a software engineer gives me the intuition that there will always be more things we'll need to change. Though I'd love it if we at some point just achieved endstate Nix which doesn't need any changes anymore and is perfect the way it is.

Reply:
> I'd love it if we at some point just achieved endstate Nix

I think if we just clean up some of the very rough edges in existing Nix we're basically there. It's got all the crucial bits already, we can soften some edges and then declare it done. I don't have any expectation of anything turning out _perfect_ though, so maybe I'm setting a lower bar ;)

Reply:
@infinisil Some of the language changes you mention cannot be handled on a per-file basis (i.e. using the tags described in this RFC). In particular, changes to `builtins` cannot be per-file since `builtins` can be passed around as an argument.

Reply:
Could different versions not get different `builtin` objects when required? This is technically breaking if you try to compare the `builtins` from two different files with different versions but I don't think that is a major issue. And in that case you can avoid updating your declared version until this issue is resolved so it won't break existing code (but does have a slight chance of making updating your eval version a breaking change)

Reply:
> In particular, changes to builtins cannot be per-file since `builtins` can be passed around as an argument.

If I pass builtins around, I already can augment them before passing, though.